### PR TITLE
apidoc: Fix bad escaping of input in conf.py_t

### PR DESCRIPTION
**Pull request by eric-wieser**
Otherwise paths like `somefolder\test` would be treated as containing a tab
character

### CODE REVIEW

**Comment by tk0miya** [comment]
Is this work with py3?
It would be nice to escape the `module_path`.

**Comment by eric-wieser** [comment]
`r` syntax is supported in all python versions.

the `u` will break in python 3.x<3.3, but that was there already, so shouldn't be an issue?

**Comment by tk0miya** [comment]
AFAIK, `ur` prefix is not supported in python3. At least, it causes SyntaxError with py34 and py36 at my local.
```
$ python3.4
Python 3.4.5 (default, Jun 27 2016, 20:10:23)
[GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> ur'hello world'
  File "<stdin>", line 1
    ur'hello world'
                  ^
SyntaxError: invalid syntax
```
```
$ python3.6
Python 3.6.1 (default, Mar 23 2017, 13:18:18)
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> ur'hello world'
  File "<stdin>", line 1
    ur'hello world'
                  ^
SyntaxError: invalid syntax
```

As a quick look, it seems dropped sicne py33 (or py34).
http://stackoverflow.com/questions/26063899/python-version-3-4-does-not-support-a-ur-prefix

**Comment by eric-wieser** [comment]
Ugh, you're right.

The goal of using `r` here was to make a path like `C:\Users\Someoné\doc` appear as `ur"C:\Users\Someoné\doc"` rather than `u"C:\\Users\\Someoné\\doc"`.

Alternatively, we could go for `br"C:\Users\Someoné\doc".decode('utf-8')`, since the file contains `# -*- coding: utf-8 -*-`?

That would look like

```python
sys.path.insert(0, br'{{ module_path.encode('utf-8') }}'.decode('utf-8'))
```

**Comment by tk0miya** [comment]
I feel `u"C:\\Users\\Someoné\\doc"` is much better.
`encode()` and `decode()` technique is too hacky for me.
Why do you want to use `r` prefix? Why not escaping?

**Comment by eric-wieser** [comment]
You actually should have the `encode` there anyway, to ensure that the encoding of the unicode matches the file encoding (unless jinja does it for you, in which case neither place needs it)

> Why do you want to use r prefix? Why not escaping?

Because then users can directly paste a file path in without worrying

> I feel `u"C:\\Users\\Someoné\\doc"` is much better.

In that case, I'd recommend

```python
sys.path.insert(0, {{ repr(module_path).encode('utf-8') }})
```

Which on python 2, gives `u'C:\\Users\\Someon\xe9\\doc'`, and on python 3, gives `'C:\\Users\\Someoné\\doc'`.


Is it important that a `conf.py` file generated with python 3 can be run on python 2 and vice versa?

**Comment by eric-wieser** [comment]
> Why do you want to use r prefix? Why not escaping?

Because then users can directly paste a file path in without worrying

**Comment by tk0miya** [comment]
I feel this is better way.
```
sys.path.insert(0, {{ repr(module_path).encode('utf-8') }})
```

>Is it important that a conf.py file generated with python 3 can be run on python 2 and vice versa?

I think this is a rare case, so it would not be problem (because py2 will be deprecated in 2020!).
@shimizukawa what do you think?

**Comment by eric-wieser** [comment]
Also worth asking - does sphinx support 2.6?

**Comment by tk0miya** [comment]
No, Sphinx supports 2.7 and 3.4+.

**Comment by shimizukawa** [comment]
IMO, the fundamental problem is that any Windows PATH specified as a option for apidoc command is render to conf.py without being escaped.
Python treats both backslashes and slashes equally as path separators, so let's use this feature. It would be nice to replace the all backslashes in the path with slash when path received from apidoc command option.

**Comment by eric-wieser** [comment]
> render to conf.py without being escaped.

Correct. And `repr` does that exact escaping - so we should use `repr`

Path separators are only a manifestation of the problem here. Presumably other strange characters could appear that need to be escaped - so we should just escape them properly.

**Comment by shimizukawa** [comment]
`repr` is not suitable for escaping backslashes. `repr` is just for representation the string.
On py2, if you include a language-specific string in the path as below, a bad string will be rendered in conf.py.

```
>>> print(repr(ur'日本語\PATH').encode('utf-8'))
u'\u65e5\u672c\u8a9e\\PATH'
>>> print(repr(ur'αβ\PATH').encode('utf-8'))
u'\u03b1\u03b2\\PATH'
>>> print(repr(ur'АБГ\PATH').encode('utf-8'))
u'\u0410\u0411\u0413\\PATH'
>>>
```

**Review by shimizukawa — CHANGES_REQUESTED** [review]
This change doesn't work with Non-ASCII characters

**Comment by eric-wieser** [comment]
> repr is not suitable for escaping backslashes. repr is just for representation the string.

This is untrue. `repr(s)` when `isinstance(s, (str, unicode))` is defined such that `eval(repr(s)) == s`. That's the definition of correct escaping. Overreaching escaping, perhaps, but definitely correct escaping.

> On py2 [...] a bad string will be rendered in conf.py.

The string that is rendered is correct. It's perhaps less readable that it could be, but it still does exactly the right thing, so "bad" is not really accurate.

For instance, `u'\u0410\u0411\u0413\\PATH' == ur'АБГ\PATH'` is `True`.

It doesn't feel like sphinx should be trying to work around a poor design choice in python 2 purely for cosmetic reasons in `conf.py`.

What does sphinx do when autodocumenting `def foo(s=ur'АБГ\PATH'): pass`? Does it render a similarly "bad"-but-correct string?

**Comment by eric-wieser** [comment]
> This change doesn't work with 

Actually, this change doesn't work at all, based on CI, as `repr` seems not to be in scope for the template...

**Comment by tk0miya** [comment]
+1 for human readable `conf.py`.
Indeed, `conf.py` is a python script, but at same time, it is a kind of configuration file. So it would be 
better if human readable.

How about replacing back-slashes with forward slashes?
I think that might work well in windows platform and human readable.

**Comment by eric-wieser** [comment]
> So it would be better if human readable.

I agree. However, here's the choice we get to make - pick one:

* Human readable on non-ascii input
* Doesn't `SyntaxError` (or worse on) input containing backslashes or quotes

I've updates my PR so that it now (hopefully) passes tests. Almost every interpolated string in this file suffered from the same bug.

In future, you could add a custom `urepr` jinja filter that emulates the python 3 `repr` behaviour on py2.

However, I don't think that needs to be in a bugfix PR.

**Comment by eric-wieser** [comment]
> How about replacing back-slashes with forward slashes?

This does not really solve the problem elsewhere.

@shimizukawa: I've added another commit with a custom repr handler that should make things both correct _and_ human readable

**Comment by eric-wieser** [comment]
> What does sphinx do when autodocumenting def foo(s=ur'АБГ\PATH'): pass? Does it render a similarly "bad"-but-correct string?

I actually think that I might have fixed this problem too here

**Comment by shimizukawa** [comment]
-1 because

- jinja2 `repr` filter implementation is too complex
- `repr` doesn't mean `backslash escaping`

(BTW shibukawa and shimizukawa is not a same person)

**Comment by eric-wieser** [comment]
@shimizukawa

> jinja2 `repr` filter implementation is too complex

The only reason it exists in the first place is to make the python 2 `repr` match the python 3 `repr`. You've already rejected the simple method of using `repr` directly in python 2 for readability reasons - which I think is reasonable.

The implementation you see here is a port of the important bits of `PyObject_Repr` from the CPython source code. Ideally it would be part of `six` or something, but it's not. So the choice is, deal with the python 2 behaviour, or invoke complexity in order to fix it.

Do you have any suggestions for how it could be made simpler?

> repr doesn't mean backslash escaping

`repr` means "take this object, and try to convert it into the python source code that generates itself". This is exactly the operation we want in a template outputting python code. From the [python 3 docs for `repr`](https://docs.python.org/3/library/functions.html#repr)

> this function makes an attempt to return a string that would yield an object with the same value when passed to `eval()`

So `repr(unicode)` absolutely **does mean backslash escaping** (and then quoting and prefixing with u)

(Apologies for confusing the usernames - I assumed github would suggest participants first in the dropdown, and didn't read the full name)

File: sphinx/util/template.py [hunk 1]
```diff
@@ -10,16 +10,65 @@
...
+        reprlib.Repr.__init__(self)
+        # disable the reprlib truncation behaviour
+        max_size = 2 ** 128
+        for k in dir(self):
+            if k.startswith('max'):
+                setattr(self, k, max_size)
+
+    if sys.version_info.major >= 3:
+        def repr_str(self, obj, level):
+            return 'u' + repr(obj)
```
Comment (eric-wieser) [inline sphinx/util/template.py]: This is new behaviour, that means that `u`s are inserted into python 3 source code. I'd argue this is useful, as it means a conf file generated in python 3 will work correctly in python 2 (for non-ascii strings).

If you want to keep things as they are, I'm happy to remove this python 3 case

**Comment by tk0miya** [comment]
It seems discussion is stopped for a long time. and @shimizukawa objected to this.
I also feel `repr` is not good for this case. So I'd like to close and file a new issue for this.

@shimizukawa what do you think?

**Comment by shimizukawa** [comment]
+1 to @tk0miya 's opinion.

**Comment by eric-wieser** [comment]
> It seems discussion is stopped for a long time

I was hoping you might respond with counterarguments to some of my points above. I don't understand why you argue `repr` is not good for this case, when this is the exact case that `repr` is designed for.

I'm happy to rebase this PR to solve conflicts if we can come to an agreement. There are also some bug fixes in here related to `unicode`/`bytes` that should go in anyway

File: sphinx/quickstart.py [hunk 2]
```diff
@@ -126,9 +125,9 @@ def term_decode(text):
     if isinstance(text, text_type):
         return text
 
-    # for Python 2.x, try to get a Unicode string out of it
+    # if ascii roundtrips, then decode as ascii
     if text.decode('ascii', 'replace').encode('ascii', 'replace') == text:
-        return text
+        return text.decode('ascii')
```
Comment (eric-wieser) [inline sphinx/quickstart.py]: This was a bug - this function should always return `unicode` (text_type)

**Comment by eric-wieser** [comment]
> implementation is too complex

I'm also planning to go back and improve the comments there based on the comments I've been making here.

**Comment by tk0miya** [comment]
>I was hoping you might respond with counterarguments to some of my points above. I don't understand why you argue repr is not good for this case, when this is the exact case that repr is designed for.

As I commented above, I think `conf.py` is used as configuration file. So readability is needed.
And I feel `u'\u0410\u0411\u0413\\PATH'` is not human readable.
I'm okay to this If your implementation does not escape Unicode character and show them as is.

In addition, I want to know why replacing back-slashes by forward-slashes is bad.

**Comment by eric-wieser** [comment]
> I'm okay to this If your implementation does not escape Unicode character and show them as is

In the latest version, this is the case. It's possibly misleading that I called the filter `repr` - perhaps calling it `nice_repr` would reduce ambiguity?

> In addition, I want to know why replacing back-slashes by forward-slashes is bad.

It's a corner case, but paths on linux can actually contain backslashes, and converting them to forward slashes would break things. Better to be correct here.

**Comment by eric-wieser** [comment]
[Demo of it working](https://www.ideone.com/NVvaHu#stdin). I think perhaps `reprlib` is doing more harm than good here, since the truncation behaviour is pretty ingrained. I'll come up with something simpler

**Comment by eric-wieser** [comment]
Ok, [here's](https://www.ideone.com/KV4ZsD) (same in [py3](https://www.ideone.com/yRpsch)) a better and simpler filter implementation, that throws out reprlib in favor of implement the few things you care about.

**Comment by eric-wieser** [comment]
I've pushed a new version, and rebased, along with some tests. `urepr` is now in its own file, to contain the complexity - I strongly suspect that there are other places in sphinx that ought to be using it.

**Comment by tk0miya** [comment]
Sorry for very late response.
I feel new `urepr` is good enough.

@shimizukawa what do you think? If you also agree, I'll start to review again.

Note: We will release 1.7 series soon. So this will be merged into `1.7-release` branch.

**Review by shimizukawa — COMMENTED** [review]
The code for urepr.py seems to be a necessary and sufficient code for what you want to realize, but I still feel that the maintenance cost is high. I will merge as soon as this PR's scope of influence only fixes path problems in Windows, but it will not be accepted as it will contain changes that affect other items in conf.py as well.

**Comment by eric-wieser** [comment]
Looks like the title of the PR is misleading - I've updated it to match the first commit message, which describes the real change. Windows paths failing is a symptom of a wider problem - incorrect data escaping.

If you'd like, I can try to write a more comprehensive explanation of the change in the commit message when I rebase to solve the conflicts.

> I still feel that the maintenance cost is high.

I'd argue the overall cost has reduced because:

* Unlike the previous code, the escaping-of-strings now is all in one place, and has unit tests.
* When python 2 is dropped, then the entire `urepr` function can be replaced with `urepr = repr`, and everything will just work.
* I suspect you need `urepr` in other places within sphinx

**Comment by tk0miya** [comment]
Sorry, I just removed `stable` branch last night. As a result, github closes this automatically. It was not intended. so I reopened this again and changed target branch to `1.7` (we decided to rename stable branch to versioned one).
Thanks,

File: sphinx/quickstart.py [hunk 3]
```diff
@@ -443,11 +425,10 @@ def generate(d, overwrite=True, silent=False, templatedir=None):
         d['exclude_patterns'] = ''
```
Comment (eric-wieser) [inline sphinx/quickstart.py:425]: Note to self - change to `[]`

**Comment by tk0miya** [comment]
I feel this PR is too longly alive, but not discussed. So I'd like to determine merging or rejecting.

In my opinion, this has some problem:

* tests have not been passed.
* @shimizukawa objects to this.
* I also feel this makes the template for conf.py complicated. +1 for using `repr` only for limited places. IMO, escaping to `dot` is too much.
* We'll drop py2 soon. So `urepr` also will be removed soon.

**Comment by tk0miya** [comment]
Now I'm closing this because this was stopped discussing for a long time.
I'm sorry for our slowly response.
Thank you for your proposal.

**Comment by eric-wieser** [comment]
> IMO, escaping to dot is too much.

What do you mean by this?

> We'll drop py2 soon. So urepr also will be removed soon

True, but you'll still need `repr` in its place

**Comment by tk0miya** [comment]
>>IMO, escaping to dot is too much.
>What do you mean by this?

This line is hard to read. I feel this is too much. Is `dot` really need to escaped?
```
templates_path = [{{ '{}templates'.format(dot) | repr }}]
```
https://github.com/sphinx-doc/sphinx/pull/3584/files#diff-b4e26a06e5a278dd53e3992672be2eedR48

>True, but you'll still need repr in its place

+1 for escaping. I don't opposite your idea itself.

**Pull request closed without merging**